### ISSUE TITLE
[compile] Modularize very long compilation

### ISSUE THREAD

**Issue by bhack** [pytorch/pytorch#147323]
### 🚀 The feature, motivation and pitch

On a model export/compile I see that there is a very very long stage (more then 1 hours) compiling a single generated c++ final file that is more that 78K+ lines

```bash
g++ /tmp/torchinductor_root/<hash>/<hash>.cpp -D TORCH_INDUCTOR_CPP_WRAPPER -D STANDALONE_TORCH_HEADER -D C10_USING_CUSTOM_GENERATED_MACROS -D CPU_CAPABILITY_AVX5+..
```

This is going to really require a lot of time, without any intermediate progress, and using only a single core at 100%.

Why isn't possible a modularization and parallelization of this monolithic generated c++ file?

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @aakhundov @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @desertfire @yushangdi @malfet @seemethere

**Comment by anijain2305** [comment]
cc @desertfire  Seems like CPP wrapper generated an extremely bad triton kernel.

@eellison I am not sure if Inductor has some guardrails to prevent itself from generating bad kernels. Not sure if this problem is specific to CPP wrapper, or just that Inductor generated a really bad kernel and Python-wrapper will suffer as well.

**Comment by bhack** [comment]
@anijain2305 Can I check something specific to support the analysis? It is a whole model export + aoti copmile and package.

**Comment by eellison** [comment]
@anijain2305 if it is one c++ final that takes forever it's likely the unrolled interpreter loop / calling convention function. and seems like a good idea to split it up. cc @benjaminglass1 who has been making a lot of good cpp wrapper compilation time improvements.

**Comment by benjaminglass1** [comment]
I've been testing building that unrolled loop at lower optimization levels, but I must admit to being...startled to hear about a 78K+ line file. I suspect that would take a long time no matter what.

@bhack Splitting into several files that can be built in parallel would certainly make sense in this case, although it would not be easy to implement in the current architecture. One potential issue is that for very small models (say, the inductor OpInfo tests), splitting into multiple files would increase build time, so we'd need some kind of heuristic for when it was worth it.

**Comment by bhack** [comment]
@eellison Can I search for some useful strings in the cpp or using pytroch logs envs to investigate the root cause?

**Comment by bhack** [comment]
I've retried with today nightly and the temp `.cpp` is `250K+` lines.

**Comment by eellison** [comment]
@bhack, if theres any abnormally large kernels, they could be a root cause.

although if it's 250K lines, maybe this is just a gigantic model. `invoke_subgraph` by @anijain2305 is intended to handle deduplicated repeated submodules / subgraphs by not inlining them and could be a good fit here.

**Comment by bhack** [comment]
I don't think it is huge. The model is https://github.com/google-deepmind/tapnet/blob/main/tapnet%2Fmodels%2Ftapir_model.py

**Comment by benjaminglass1** [comment]
@eellison I'm planning to dig into this some on Monday, but I don't think `cpp_wrapper` supports subgraph handling currently, so if this _is_ somewhere that that would be useful, that might be part of why this is so big.

**Comment by bhack** [comment]
You could need to make a few  very minimal changes to make it `export`+`aoti compile` compatible.

**Comment by benjaminglass1** [comment]
@bhack I haven't had a chance to look into this this week, but I would be interested in whether you see substantive compile time improvements when setting `AOT_INDUCTOR_COMPILE_WRAPPER_WITH_O0=1` in your environment while running this model. One area of active work I'm pursuing is trying to only build performance critical parts of the model with `-O3` optimization, leaving the rest at `-O1`, and using this setting may give us an idea of whether that would help you out at all.

**Comment by bhack** [comment]
I had another 57K generate cpp on a new model and `AOT_INDUCTOR_COMPILE_WRAPPER_WITH_O0=1` it seems to not help. It is already 40+ mins on 7 core (but the cpp part is single core of course).

**Comment by bhack** [comment]
In any case with this amount of compile time and limited on single core it is impossible to iterate over an export + aot-inference development.

**Comment by benjaminglass1** [comment]
@bhack Thank you for investigating! I'm planning to look at this issue this week, so hopefully we'll have some answers on how to make this better soon.

**Comment by bhack** [comment]
Same, without `O0` ENV is still compiling at 180 mins (as it is a unique cpp single core).

If you want to check the model (you need to trick the dynamism in some model functions to export it):
https://github.com/rjsaurav13/TrackOn

**Comment by bhack** [comment]
@benjaminglass1 On this last specific case it reached 160K cpp lines.

**Comment by bhack** [comment]
@benjaminglass1 You are not able to reproduce cause the aoti compile is failing?

**Comment by benjaminglass1** [comment]
@bhack No, I worked around the dynamic shapes issue fairly easily, but the output file was on the order of 12k lines rather than 70k. I was roughly following what was done in their example PyTorch [colab notebook](https://colab.research.google.com/github/deepmind/tapnet/blob/master/colabs/torch_tapir_demo.ipynb).

**Comment by bhack** [comment]
@benjaminglass1  With pytorch nightly (cuda 12.6). I am using the official nightly docker image (last time I've tested was `2.7.0.dev20250214-cuda12.6-cudnn9-devel`).

Can you try this with and without `dynamc_shapes`
```python
import torch
from tapnet.torch import tapir_model
from torch._inductor import aoti_compile_and_package
from tapnet.torch.utils import generate_default_resolutions
from torch.export import Dim

# Minimal repro
checkpoint_path = "<checkpoint-path>"
export_path = "<export-path>"
device = torch.device("cuda")

model = tapir_model.TAPIR(pyramid_level=1).to(device).eval()
model.load_state_dict(torch.load(checkpoint_path))

H, W = 256, 256
use_amp = True
input_dtype = torch.bfloat16 if use_amp and torch.cuda.is_bf16_supported() else torch.float32

frames = torch.randn(1, 10, H, W, 3, device=device, dtype=input_dtype)
query_points = torch.rand(1, 50, 3, device=device)
refinement_resolutions = generate_default_resolutions(full_size=(H, W), train_size=(256, 256))

# Dynamic shapes option
dynamic_shapes = {
    "video": {1: Dim.DYNAMIC},
    "query_points": {0: Dim.STATIC}
}

exported_program = torch.export.export(
    model,
    (frames, query_points),
    kwargs={"refinement_resolutions": refinement_resolutions},
    dynamic_shapes=dynamic_shapes
)

aoti_compile_and_package(exported_program, package_path=export_path, inductor_configs={"max_autotune": True})
```

**Comment by bhack** [comment]
P.s. I was using `BootsTAPIR` checkpoints https://github.com/google-deepmind/tapnet?tab=readme-ov-file#checkpoints

**Comment by benjaminglass1** [comment]
@bhack Much thanks for the reproducer! I'll dig into this.

**Comment by benjaminglass1** [comment]
@bhack After a bit of struggle getting everything to compile in the first place (I had to make some tweaks to your test script, plus some edits to the TAPIR model to get everything to compile), I believe I have some excellent news for you! If you build with the changes in #148305 (recently merged, then reverted due to broken ROCm tests), the generated file is only about 12k lines, and builds fairly quickly (30 seconds or so). Examining the output code, that makes a lot of sense; prior to that PR, Triton kernel calls and associated variable declarations were repeated every time the kernel needed to be called, but now all that is put inside a single function definition that's repeatedly called instead. This greatly reduces the number of variables and number of lines of code, effectively solving your problem.

**Comment by bhack** [comment]
> After a bit of struggle getting everything to compile in the first place (I had to make some tweaks to your test script, plus some edits to the TAPIR model to get everything to compile),

Unfortunately, there is always a lot of work to do on every model with the current compile/export maturity. Do you think is the same case also for https://github.com/rjsaurav13/TrackOn has it generated 160K lines.

**Comment by benjaminglass1** [comment]
@bhack I'm not sure what model you're running from that repo, sorry. It looks to be written in C? However, I'd expect that it would dramatically fall in line count, unless it's an incredibly complicated model in its own right.

WRT the compile/export maturity, please keep opening issues as you run into things! This one was a helpful jumping off point for investigation. You could also consider running Inductor in `cpp_wrapper` mode, which is more broadly compatible now than AOT export, but still involves the Python runtime at times (so it's slower).

**Comment by bhack** [comment]
Sorry my bad same name wrong link. The model is  https://github.com/gorkaydemir/track_on

**Comment by bhack** [comment]
If you can give a try you will see that these memory models really stress the entire stack. I've exported with a similar approach approach as TAPIR but it is generating more then 160K lines. 
Also it is quite tricky for the stack I was able to export it in a single step version that it is going to require to re-implement a lot of state passing logic on aoti inference.

Other then the amount of work it is going to partially neutralize the advantage of an export+aoti integration.

**Comment by benjaminglass1** [comment]
@bhack Spent the better part of this afternoon trying to get this model to export, but kept running into some deep-in-the-stack errors with libraries using C-extension functions export couldn't understand. I'm not entirely sure how you worked around that in your testing, but if you have a local copy of `track_on` setup and working, I'd recommend trying the PyTorch nightly when it next uploads, and checking in on the generated code length. I suspect you'll see substantive improvements.

**Comment by bhack** [comment]
@benjaminglass1 You see how much is hard on real life when you are not working with very static models and eventually with custo ops. 
Of course I always test with today nightly before reporting bugs.

> C-extension functions export
Exactly, but probably you know how much is popular to reuse encoders or ops from different popular libraries in research.
In this specific case as we have never solved a quite requested ops like https://github.com/pytorch/pytorch/issues/112827  researchers rely on third party implementation. In this specific case they are using the popular `mmcv` just for the op (https://mmcv.readthedocs.io/en/latest/_modules/mmcv/ops/multi_scale_deform_attn.html),

So just to start to reason about compiling/export this we need to hotpatch that API call preparing what it is currently required from the doc:
https://pytorch.org/tutorials/intermediate/torch_export_tutorial.html#custom-ops

Then we start to dance with all the loop unrolling issue that enforce things to be static related just on the dummy input for export as on aoti export we cannot compile multiple dummy inputs like in "online" torch.compile. And we open the door to all "classical "issues that we have in the current stack.

**Comment by benjaminglass1** [comment]
@bhack The recommendation about trying the nightly was solely because the PR that may fix this issue merged today. On the rest, I hear you. 👍

**Comment by bhack** [comment]
Yes but this is just to give the evidence how much work it needs to be shared just to let you test a reproduction.

As with the current status of the stack it could require some day of work and work arounds before the bug just for the reproduction.

**Comment by benjaminglass1** [comment]
@bhack I'm going to go ahead and mark this issue as closed, but feel free to open a new one if the merged PR I pointed you to doesn't fix things.
